xarray/coding/variables.py

```diff
@@ -188,7 +189,89 @@ def decode(self, variable, name=None):
             if encoded_fill_values:
                 transform = partial(_apply_mask,
                                     encoded_fill_values=encoded_fill_values,
-                                    decoded_fill_value=decoded_fill_value)
+                                    decoded_fill_value=decoded_fill_value,
+                                    dtype=dtype)
                 data = lazy_elemwise_func(data, transform, dtype)
 
         return Variable(dims, data, attrs, encoding)
+
+
+def _scale_offset_decoding(data, scale_factor, add_offset, dtype):
+    data = np.array(data, dtype=dtype, copy=True)
+    if scale_factor is not None:
+        data *= scale_factor
+    if add_offset is not None:
+        data += add_offset
+    return data
+
+
+class CFScaleOffsetCoder(VariableCoder):
+    """Scale and offset variables according to CF conventions.
+
+    Follows the formula:
+        decode_values = encoded_values * scale_factor + add_offset
+    """
+
+    def encode(self, variable, name=None):
+        dims, data, attrs, encoding = unpack_for_encoding(variable)
+
+        if 'scale_factor' in encoding or 'add_offset' in encoding:
+            data = data.astype(dtype=np.float64, copy=True)
+            if 'add_offset' in encoding:
+                data -= pop_to(encoding, attrs, 'add_offset', name=name)
+            if 'scale_factor' in encoding:
+                data /= pop_to(encoding, attrs, 'scale_factor', name=name)
+
+        return Variable(dims, data, attrs, encoding)
+
+    def decode(self, variable, name=None):
+        dims, data, attrs, encoding = unpack_for_decoding(variable)
+
+        if 'scale_factor' in attrs or 'add_offset' in attrs:
+            scale_factor = pop_to(attrs, encoding, 'scale_factor', name=name)
+            add_offset = pop_to(attrs, encoding, 'add_offset', name=name)
+            dtype = np.float64
+            transform = partial(_scale_offset_decoding,
+                                scale_factor=scale_factor,
+                                add_offset=add_offset,
+                                dtype=dtype)
+            data = lazy_elemwise_func(data, transform, dtype)
```

Comment:
This looks great! When did `lazy_elemwise_func` get added? I missed that somehow. This solves our problems related to multiple types of lazy arrays.

Reply:
@rabernat - see https://github.com/pydata/xarray/pull/1752